awx/main/tasks.py

```diff
@@ -2327,6 +2328,19 @@ def pre_run_hook(self, instance, private_data_dir):
                 else:
                     self.original_branch = git_repo.active_branch
 
+    def clear_project_cache(self, instance, revision):
+        cache_dir = instance.get_cache_path()
+        for entry in os.listdir(cache_dir):
+            old_path = os.path.join(cache_dir, entry)
+            if entry != revision:
+                # invalidate, then delete
+                new_path = os.path.join(cache_dir,'.~~delete~~' + entry)
+                try:
+                    os.rename(old_path, new_path)
+                    shutil.rmtree(new_path)
```

Comment:
this should be in a background thread, if we can stomach the complexity (we're inside the update lock here, and this may be a long deletion)